design/node-agent-concurrency.md

```diff
@@ -0,0 +1,77 @@
+- Help users to isolate critical environments. Users may run some critical workloads in some specified nodes, since VGDP instances may take large resource consumption, users may only want to run them in the other nodes or run less number of instances in the nodes with critical workloads
+
+The range of Per-node concurrent number is the same with Global concurrent number.  
+Per-node concurrent number is preferable to Global concurrent number, so it will overwrite the Global concurrent number for that node.  
+Per-node concurrent number is implemented in a separate configMap named ```node-agent-configs```. This configMap is not created by Velero, users should create it manually on demand. The configMap should be created before node-agent server starts, otherwise, node-agent server should be restarted in order to make the effect.  
+
+The ```node-agent-configs``` configMap may be used for other purpose of configuring node-agent in future, at present, there is only one kind of data ```data-path-concurrency```. Its data struct is like below: 
+
+```go
+type DataPathConcurrency struct {
+	// MatchLabel specifies the label to identify a node for which Configs holds configs
+	MatchLabel string `json:"matchLabel"`
+
+	// Configs specifies the configs by node identified by MatchLabel
+	Configs map[string]int `json:"configs"`
+}
+```
+
+Users are allowed to specify how the node is matched. The ```MatchLabel``` field accepts the key of a label associated to nodes. Then this label is used to define concurrent number in ```Configs``` field.  
+
+For example, say ```MatchLabel``` is "foo", then:  
+- At least one node is expected to have a label with key of "foo" (e.g., "foo":"bar"). If no node is with this label, the Per-node configuration makes no effect  
+- Inside the map of ```Configs```, if there is one element "bar":X, then X will be the Per-node concurrent number for the node with the label "foo":"bar"  
+
+```MatchLabel``` is optional, if it is not set, the hard-coded label key ```kubernetes.io/hostname``` will be used, which means the host name will be used to match nodes.  
+
+### Global data path manager
+As for the code implementation, data path manager is to maintain the total number of the running VGDP instances and ensure the limit is not excceeded. At present, there is one data path manager instance per controller, as a result, the concurrent numbers are calculated separately for each controller. This doesn't help to limit the concurrency among different requesters.  
+Therefore, we need to create one global data path manager instance server-wide, and pass it to different controllers. The instance will be created at node-agent server startup.  
+The concurrent number is required to initiate a data path manager, the number comes from either Per-node concurrent number or Global concurrent number.    
```

Comment:
How do we control the limit of the running VGDP instances? Is this already implemented by Velero or Kopia?

Reply:
Yes, it has been implemented, check the [dataPath manager code](https://github.com/vmware-tanzu/velero/blob/main/pkg/datapath/manager.go)

Reply:
Is it possible to also update the PVB, PVR, DataUpload, and DataDownload status to reflect the reason why they are still not running?

```golang
	fsBackup, err := r.dataPathMgr.CreateFileSystemBR(pvb.Name, pVBRRequestor, ctx, r.Client, pvb.Namespace, callbacks, log)
	if err != nil {
		if err == datapath.ConcurrentLimitExceed {
			return ctrl.Result{Requeue: true, RequeueAfter: time.Minute}, nil
		} else {
			return r.errorOut(ctx, &pvb, err, "error to create data path", log)
		}
	}
```

Reply:
OK, will do this in the code change PR.